feat(profiling): integrate scoped-based timer to onedim core

tai271828 opened the pull request:
This pull request is the first step of #135 .

# How to test this pull request
Run this driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py). The following output will be generated on your stdout after all pictures of shock tube are played and the image window is closed.

```
Euler1DCore.march_alpha2 : count = 51 , time = 0.0365707 (second)
Euler1DCore.setup_march : count = 1 , time = 2.2676e-05 (second)
Euler1DCore::density : count = 51 , time = 0.0020337 (second)
Euler1DCore::initialize_data : count = 1 , time = 1.6329e-05 (second)
Euler1DCore::march_half_so0 : count = 102 , time = 0.0179795 (second)
Euler1DCore::pressure : count = 51 , time = 0.00219614 (second)
Euler1DCore::update_cfl : count = 103 , time = 0.00225837 (second)
Euler1DCore::velocity : count = 51 , time = 0.00142665 (second)
Euler1DKernel::derive : count = 20298 , time = 0.0137601 (second)
Euler1DKernel::march_half_so1_alpha : count = 102 , time = 0.01569 (second)
```

tai271828 committed:
feat(profiling): integrate scoped-based timer to onedim core

yungyuc commented on cpp/modmesh/onedim/Euler1DCore.cpp:
```diff
@@ -111,6 +117,7 @@ void Euler1DCore::update_cfl(bool odd_plane)
 
 void Euler1DCore::march_half_so0(bool odd_plane)
```
Could you please also add the profiling hooks to the first-order derivative calculating functions (templates)?

tai271828 replied:
Will do!

tai271828 replied:
The gist has been appended in the original description.

tai271828 replied:
Completed and updated this PR.

yungyuc reviewed (changes requested):
Could you also add the profiling driver in a gist and include it in the PR for reference?

tai271828 committed:
feat(profiling): include necessary hearder only

The previous includes toggle.hpp but we only need MODMESH_TIME in profile.hpp

tai271828 committed:
gitignore: ignore dot env files

Dot env files are often used for management of Python environment
variable. Many IDEs and tools support to use it to set up runtime and
debugging environment of Python. Ignoring dot env files would help us
not commit different dot env files for different kinds of runtime and
debugging environment setup.

tai271828 committed:
feat(profiling): profile Euler1DKernel::derive

tai271828 committed:
feat(profiling): profile Euler1DKernel::march_half_so1_alpha

tai271828 commented:
I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.

yungyuc commented:
> I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.

What about integrating the script `euler1d_profiling.py` to the euler1d app and then simplifying the driving script to be a one-liner?  The one-liner can be mentioned in the comment in the app.

tai271828 committed:
feat(app): provide euler1d with profiling option

With this option, users can choose if they want to see profiling data
when using REPL of viewer.

tai271828 commented:
> > I prepared a new driving script from gist [euler1d_profiling-py](https://gist.github.com/tai271828/d5f50e4a06274b1203ed79f6739039dd#file-euler1d_profiling-py) for testing (and also updated the corresponding description session above). This driving script runs through a real shock tube problem and the result is easier to ready.
> 
> What about integrating the script `euler1d_profiling.py` to the euler1d app and then simplifying the driving script to be a one-liner? The one-liner can be mentioned in the comment in the app.

Sounds good. Let me work on it.

tai271828 committed:
feat(app): make euler1d easier for being tested

This commit makes euler1d.py could be executed directly by a Python
interpreter without launching a modmesh viewer. By introducing this
feature, we can integrate this app into CI easily.

Usage: python ./modmesh/app/euler1d.py

tai271828 committed:
feat(app): hint of euler1d for non-PySide6 users

Commit:
style(app): fix flake8 error for euler1d

tai271828 commented on modmesh/app/euler1d.py:
```diff
@@ -25,6 +25,9 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 
+__package__ = "modmesh.app"
```
A trick to make euler.py runnable in command line.

tai271828 replied:
@yungyuc thanks for adding the comments. This one is really tricky so I am also very +1 to adding comments on it. Thanks for the update!

tai271828 commented on modmesh/app/euler1d.py:
```diff
@@ -46,6 +49,9 @@ def load_app():
 ctrl.step()  # Stepping the solution
+
+# Note: you can enable profiling info by "profiling" option
+ctrl = mm.app.euler1d.run(interval=10, max_steps=50, profiling=True)
```
The one-liner aforementioned https://github.com/solvcon/modmesh/pull/190#issuecomment-1383108362 .

yungyuc replied:
This is great!

tai271828 commented on modmesh/app/euler1d.py:
```diff
@@ -175,16 +185,56 @@ def log(msg):
 
-    ctrl = Controller(shocktube=st, max_steps=max_steps, **kw)
+    if no_view_mgr:
+        ctrl = ControllerNoViewMgr(shocktube=st, max_steps=max_steps, **kw)
```
When running in a terminal (for automation), we don't launch view manager.

yungyuc committed:
style(app): add comment
